graphql_compiler/compiler/emit_cypher.py

```diff
@@ -127,10 +159,120 @@ def _emit_with_clause_components(cypher_steps):
     return result
 
 
+def _emit_with_clause_components_for_current_fold_scope(current_fold_scope_cypher_steps):
+    """Emit a list of strings, one for each vertex or list passing through a WITH clause."""
+    result = []
+
+    vertex_names = {}
+    for cypher_step in current_fold_scope_cypher_steps:
+        fold_scope_location = cypher_step.as_block.location
+        full_vertex_name = fold_scope_location_full_path_name(fold_scope_location)
+        collected_name = collected_vertex_list_name(full_vertex_name)
+        vertex_names[u'collect(' + full_vertex_name + ')'] = collected_name
+
+    # Sort the locations, to ensure a deterministic order.
+    for index, collect_call in enumerate(sorted(vertex_names)):
+        if index > 0:
+            result.append(u',')
+        result.append(u'\n  %(collect_call)s AS %(collected_name)s' %
+                      {'collect_call': collect_call, 'collected_name': vertex_names[collect_call]})
+    return result
+
+
+def _emit_fold_scope(cypher_query):
+    """Return a Cypher query pattern match expression for each fold scope in cypher_query.
+
+    Consider the following CYPHER query, corresponding to test_input_data.multiple_folds()
+
+    MATCH (Animal___1:Animal)
+    OPTIONAL MATCH (Animal___1)-[:Animal_ParentOf]->(Animal__out_Animal_ParentOf___1:Animal)
+    WITH
+      Animal___1 AS Animal___1,
+      collect(Animal__out_Animal_ParentOf___1) AS collected_Animal__out_Animal_ParentOf___1
+    OPTIONAL MATCH (Animal___1)<-[:Animal_ParentOf]-(Animal__in_Animal_ParentOf___1:Animal)
+    WITH
+      Animal___1 AS Animal___1,
+      collected_Animal__out_Animal_ParentOf___1 AS
+        collected_Animal__out_Animal_ParentOf___1,
+      collect(Animal__in_Animal_ParentOf___1) AS
+        collected_Animal__in_Animal_ParentOf___1
+    RETURN
+      Animal___1.name AS `animal_name`,
+      [x IN collected_Animal__out_Animal_ParentOf___1 | x.name] AS `child_names_list`,
+      [x IN collected_Animal__out_Animal_ParentOf___1 | x.uuid] AS `child_uuids_list`,
+      [x IN collected_Animal__in_Animal_ParentOf___1 | x.name] AS `parent_names_list`,
+      [x IN collected_Animal__in_Animal_ParentOf___1 | x.uuid] AS `parent_uuids_list`
+
+    For each fold scope, the structure generally works as follows:
+
+    First, traverse to the vertex just outside the fold scope. Each traversal is mandatory here, so
```

Comment:
For the enumeration that starts on line 206, introduce some sort of bullet point symbol so that it's clear where each bullet point starts and ends. I'd suggest `-`.